UnrealPlugin/Source/PlayfabGSDK/Private/FGSDKInternal.cpp

```diff
@@ -0,0 +1,374 @@
+
+		TSharedPtr<FJsonObject> SessionConfigJson = HeartbeatResponseJson->GetObjectField(TEXT("sessionConfig"));
+		for (const auto& SessionConfigJsonValue: SessionConfigJson->Values)
+		{
+			FString ValueString;
+			if (SessionConfigJsonValue.Value->TryGetString(ValueString))
+			{
+				if (ConfigSettings.Contains(SessionConfigJsonValue.Key))
+				{
+					ConfigSettings[SessionConfigJsonValue.Key] = ValueString;
+				}
+				else
+				{
+					ConfigSettings.Add(SessionConfigJsonValue.Key, ValueString);
+				}
+			}
+		}		
+
+		// Update initial players only if this is the first time populating it.
+		if (InitialPlayers.Num() == 0 && SessionConfigJson->HasField(TEXT("initialPlayers")))
+		{
+			TArray<TSharedPtr<FJsonValue>> PlayersJson = SessionConfigJson->GetArrayField(TEXT("initialPlayers"));
+
+			for (auto PlayerJson: PlayersJson)
+			{
+				InitialPlayers.Add(PlayerJson->AsString());
+			}
+		}
+
+		if (SessionConfigJson->HasField(TEXT("metadata")))
+		{
+			TSharedPtr<FJsonObject> MetaDatasJson = SessionConfigJson->GetObjectField(TEXT("metadata"));
+
+			for (const auto& MetaDataJson: MetaDatasJson->Values)
+			{
+				FString ValueString;
+				if (MetaDataJson.Value->TryGetString(ValueString))
+				{
+					if (ConfigSettings.Contains(MetaDataJson.Key))
+					{
+						ConfigSettings[MetaDataJson.Key] = ValueString;
+					}
+					else
+					{
+						ConfigSettings.Add(MetaDataJson.Key, ValueString);
+					}
+				}
+			}
+		}
+	}
+
+	if (HeartbeatResponseJson->HasField("nextScheduledMaintenanceUtc"))
+	{
+		FDateTime NextMaintenance = ParseDate(HeartbeatResponseJson->GetStringField(TEXT("nextScheduledMaintenanceUtc")));
+		FTimespan Diff = NextMaintenance - CachedScheduledMaintenance;
+
+		if (OnMaintenance.IsBound() && (!Diff.IsZero() || CachedScheduledMaintenance.GetTicks()))
+		{
+			OnMaintenance.Execute(NextMaintenance);
+			CachedScheduledMaintenance = NextMaintenance;
+		}
+	}
+
+	if (HeartbeatResponseJson->HasField(TEXT("operation")))
+	{
+		EOperation NextOperation = OperationMap[HeartbeatResponseJson->GetStringField(TEXT("operation"))];
+
+		switch (NextOperation)
+		{
+		case EOperation::Continue:
+			{
+				break;
+			}
+		case EOperation::Active:
+			{
+				if (HeartbeatRequest.CurrentGameState != EGameState::Active)
+				{
+					SetState(EGameState::Active);
+					TransitionToActiveEvent->Trigger();
+				}
+				break;
+			}
+		case EOperation::Terminate:
+			{
+				if (HeartbeatRequest.CurrentGameState != EGameState::Terminating)
+				{
+					SetState(EGameState::Terminating);
+					TransitionToActiveEvent->Trigger();
+					AsyncTask(ENamedThreads::AnyThread, [this]()
+					{
+						if (this->OnShutdown.IsBound())
+						{
+							this->OnShutdown.Execute();
+						}
+
+						this->KeepHeartbeatRunning = false;
+					});
+				}
+				break;
+			}
+		default:
+			{
+				UE_LOG(LogPlayfabGSDK, Error, TEXT("Unhandled operation received: %s"), OperationNames[static_cast<int32>(NextOperation)]);
+				break;
```

Comment:
I think I would count the number of unexpected operations, and if that count ever exceeds a certain number in a row, like 10, or 100, or even better, a value pulled from config, then it does a shutdown. Basically if you spend an extended amount of time getting nothing but garbage, then it's likely you need to shut down.
Probably reset to 0 for every valid operation processed.

Reply:
Alternate algorithms could be:
subtract 2, divide by 2, sqrt, etc. Ultimately all of those boil MOSTLY down to a fixed number anyway, but just makes it harder to ultimately decide what that number really is.
I think they're all ultimately nearly identical in that they boil down to a single number of max errors, but reset to 0 has a much clearer precise number of errors-in-a-row that are allowed vs not.

Reply:
I can add a counter, but if the case would happen, doesn't it mean that the plugin would require an update, because there was maybe a new state introduced, that the plugin doesn't support, so therefore, whoever uses the plugin, would constantly crash, if this new state happens multiple times.